sandbox/lua/modules/line_protocol.lua

```diff
@@ -0,0 +1,207 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[=[
+Provides functions to convert full Heka message contents to line protocol for
+InfluxDB HTTP write API (new in InfluxDB v0.9.0) or Graphite/Carbon.
+Optionally includes all standard message fields as tags or fields and
+iterates through all of the dynamic fields
+to add as points (series), skipping any fields explicitly omitted using the
+`skip_fields` config option.  It can also map any Heka message fields as tags
+in the request sent to the InfluxDB write API, using the `tag_fields` config
+option.  All dynamic fields in the Heka message are converted to separate points
+separated by newlines that are submitted to InfluxDB.
+
+API
+^^^
+
+**carbon_line_msg()**
+    Wrapper function that calls others within this module and the field_util
+    module to generate a table of Carbon line protocol messages that are
+    derived from the dynamic fields in a Heka message. Base fields or
+    dynamic fields can be used in the metric name portion of the message.
+    Configuration is implemented in the encoders that utilize this module.
+    This function takes no arguments.
+
+**influxdb_line_msg()**
+    Wrapper function that calls others within this module and the field_util
+    module to generate a table of InfluxDB line protocol messages that are
+    derived from the base or dynamic fields in a Heka message. Base fields or
+    dynamic fields can be used in the metric name portion of the message and
+    included as tags if desired.  Configuration is implemented in the encoders
+    that utilize this module.  This function takes no arguments.
+
+--]=]
+
+local field_util = require "field_util"
+local math = require "math"
+local read_config = read_config
+local read_message = read_message
+local read_next_field = read_next_field
+local pairs = pairs
+local string = require "string"
+local table = require "table"
+local tostring = tostring
+local type = type
+
+local M = {}
+setfenv(1, M) -- Remove external access to contain everything in the module.
+
+if not timestamp_precision then
+    timestamp_precision = "ms"
+end
+if not decimal_precision then
+    decimal_precision = "6"
+end
+if not name_prefix then
+    name_prefix = ""
+end
+if not name_prefix_delimiter then
+    name_prefix_delimiter = ""
+end
+if not value_field_name then
+    value_field_name = "value"
+end
+
+-- Remove blacklisted fields from the set of base fields that we use, and
+-- create a table of dynamic fields to skip.
+local skip_fields_str = read_config("skip_fields") or ""
+local skip_fields, skip_fields_all_base = field_util.field_map(skip_fields_str)
+local used_base_fields = field_util.used_base_fields(skip_fields)
+
+-- Create and populate a table of fields to be used as tags
+local tag_fields_str = read_config("tag_fields") or "**all_base**"
```

Comment:
Okay, I think I see why you did it this way, since you're using these values in the `used_base_fields` and `field_map` calls happening at module scope, and you need to make sure you have the config values even before they might be set by the client code. I'm still a bit wary of it, though.

Maybe instead of having the client code set attributes on the `line_protocol` module, we can have a `line_protocol_config` function that the client code must call before calling any other functions in the module, passing the needed values in as separate arguments, or maybe embedded in a Lua table. That smells a bit better than setting module-level attributes from outside, even if internally that's still what we're doing. It would also mean that all of the `read_config` calls can be moved to the encoders.